clang/runtime/dpct-rt/include/dpct/sparse_utils.hpp

```diff
@@ -40,13 +40,193 @@ class matrix_info {
+  inline static rel_func_t _rel_func = nullptr;
+  std::vector<sycl::event> _deps;
+};
+template <>
+inline handle_manager<oneapi::mkl::sparse::matrix_handle_t>::init_func_t
+    handle_manager<oneapi::mkl::sparse::matrix_handle_t>::_init_func =
+        oneapi::mkl::sparse::init_matrix_handle;
+template <>
+inline handle_manager<oneapi::mkl::sparse::matrix_handle_t>::rel_func_t
+    handle_manager<oneapi::mkl::sparse::matrix_handle_t>::_rel_func =
+        oneapi::mkl::sparse::release_matrix_handle;
+
+template <>
+inline handle_manager<oneapi::mkl::sparse::matmat_descr_t>::init_func_t
+    handle_manager<oneapi::mkl::sparse::matmat_descr_t>::_init_func =
+        oneapi::mkl::sparse::init_matmat_descr;
+template <>
+inline handle_manager<oneapi::mkl::sparse::matmat_descr_t>::rel_func_t
+    handle_manager<oneapi::mkl::sparse::matmat_descr_t>::_rel_func =
+        [](sycl::queue &queue, oneapi::mkl::sparse::matmat_descr_t *p_desc,
+           const std::vector<sycl::event> &dependencies) -> sycl::event {
+  oneapi::mkl::sparse::release_matmat_descr(p_desc);
+  return queue.submit([&](sycl::handler &cgh) {
+    cgh.depends_on(dependencies);
+    cgh.host_task([] {});
+  });
+};
+class matrix_handle_manager
+    : public handle_manager<oneapi::mkl::sparse::matrix_handle_t> {
+public:
+  matrix_handle_manager()
+      : handle_manager<oneapi::mkl::sparse::matrix_handle_t>() {}
+  matrix_handle_manager(sycl::queue *q)
+      : handle_manager<oneapi::mkl::sparse::matrix_handle_t>(q) {}
+  matrix_handle_manager(const handle_manager &other) = delete;
+  matrix_handle_manager operator=(handle_manager other) = delete;
+
+private:
+#ifdef DPCT_USM_LEVEL_NONE
+  static constexpr size_t _max_data_variable_size =
+      (std::max)({sizeof(sycl::buffer<float>), sizeof(sycl::buffer<double>),
+                  sizeof(sycl::buffer<std::complex<float>>),
+                  sizeof(sycl::buffer<std::complex<double>>)});
+  using value_buf_t =
+      std::variant<std::array<std::byte, _max_data_variable_size>,
+                   sycl::buffer<float>, sycl::buffer<double>,
+                   sycl::buffer<std::complex<float>>,
+                   sycl::buffer<std::complex<double>>>;
+  sycl::buffer<int> _row_ptr_buf = sycl::buffer<int>(0);
+  sycl::buffer<int> _col_ind_buf = sycl::buffer<int>(0);
+  value_buf_t _val_buf;
+  template <typename T, bool placement_new>
+  friend void set_matrix_data(matrix_handle_manager *memory, sycl::queue &queue,
+                              const int rows, const int cols,
+                              oneapi::mkl::index_base base, int *row_ptr,
+                              int *col_ind, T *val);
+#endif
+};
+template <typename Ty, bool placement_new = true>
+void set_matrix_data(matrix_handle_manager *memory, sycl::queue &queue,
+                     const int rows, const int cols,
+                     oneapi::mkl::index_base base, int *row_ptr, int *col_ind,
+                     Ty *val) {
```

Comment:
```suggestion
                     oneapi::mkl::index_base base, void *row_ptr, void *col_ind,
                     void* val) {
```
cast in the function